### PR TITLE
feat: improve subscriptions

### PR DESCRIPTION
**Pull request by jensneuse**

### CODE REVIEW

File: router/core/router.go [hunk 1]
```diff
@@ -799,6 +799,12 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
 		MaxOperationSizeInBytes: int64(r.routerTrafficConfig.MaxRequestBodyBytes),
 		PersistentOpClient:      r.cdnPersistentOpClient,
 	})
+	// Pre-hash all data source IDs to avoid races
```
Comment (StarpTech) [inline router/core/router.go:802]: Could you add a few more comments why we need to add this here? For me it's also a TODO to do it in a different place at the engine side.

File: router/core/websocket_stats.go [hunk 2]
```diff
@@ -75,12 +79,14 @@ func (s *WebSocketStats) run(ctx context.Context) {
 			s.reportConnections()
 		case <-s.update:
 			s.mu.Lock()
-			for _, subscriber := range s.subscribers {
+			for ctx, subscriber := range s.subscribers {
 				// non-blocking send
```
Comment (StarpTech) [inline router/core/websocket_stats.go:83]: This comment is no longer valid.

File: router/internal/epoller/epoll_linux.go [hunk 3]
```diff
@@ -99,7 +103,7 @@ func (e *Epoll) Wait(count int) ([]net.Conn, error) {
 	events := make([]unix.EpollEvent, count)
 
 retry:
-	n, err := unix.EpollWait(e.fd, events, -1)
+	n, err := unix.EpollWait(e.fd, events, e.timeoutMsec)
```
Comment (StarpTech) [inline router/internal/epoller/epoll_linux.go:106]: Did you find out why -1 is used on linux? Without any good argument I would not change the default.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/internal/epoller/epoll_linux.go (hunk 3 above)
Comment (jensneuse) [reply router/internal/epoller/epoll_linux.go:106]: -1 = no timeout

File: router/core/router.go (hunk 1 above)
Comment (jensneuse) [reply router/core/router.go:802]: done

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**